// @vk-io/streaming/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vkIo = require('vk-io');
var WebSocket = require('ws');
var fetch = require('node-fetch');
var createDebug = require('debug');
var inspectable = require('inspectable');
var https = require('https');
var url = require('url');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var WebSocket__default = /*#__PURE__*/_interopDefaultLegacy(WebSocket);
var fetch__default = /*#__PURE__*/_interopDefaultLegacy(fetch);
var createDebug__default = /*#__PURE__*/_interopDefaultLegacy(createDebug);

class StreamingRuleError extends vkIo.VKError {
    /**
     * Constructor
     */
    constructor({ message, error_code: code }) {
        super({ message, code });
    }
}

/**
 * Copies object params to new object
 */
const copyParams = (params, properties) => {
    const copies = {};
    for (const property of properties) {
        copies[property] = params[property];
    }
    return copies;
};

class StreamingContext extends vkIo.Context {
    constructor(options) {
        const { action, event_type: type } = options.payload;
        super({
            ...options,
            type: 'publication',
            subTypes: [
                `publication_${type}`,
                `${action}_publication`,
                `${action}_publication_${type}`
            ]
        });
        this.attachments = vkIo.transformAttachments(this.payload.attachments, this.api);
    }
    /**
     * Checks is new object
     */
    get isNew() {
        return this.actionType === 'new';
    }
    /**
     * Checks is update object
     */
    get isUpdate() {
        return this.actionType === 'update';
    }
    /**
     * Checks is delete object
     */
    get isDelete() {
        return this.actionType === 'delete';
    }
    /**
     * Checks is restore object
     */
    get isRestore() {
        return this.actionType === 'restore';
    }
    /**
     * Checks is post event
     */
    get isPost() {
        return this.eventType === 'post';
    }
    /**
     * Checks is share event
     */
    get isShare() {
        return this.eventType === 'share';
    }
    /**
     * Checks is comment event
     */
    get isComment() {
        return this.eventType === 'comment';
    }
    /**
     * Returns the event URL
     */
    get url() {
        return this.payload.event_url;
    }
    /**
     * Returns the creation time
     */
    get createdAt() {
        return this.payload.creation_time;
    }
    /**
     * Returns the text of the post
     */
    get text() {
        return this.payload.text;
    }
    /**
     * Returns the text of the shared post
     */
    get sharedText() {
        return this.payload.shared_post_text;
    }
    /**
     * Returns the creation time from original post
     */
    get sharedAt() {
        return this.payload.shared_post_creation_time;
    }
    /**
     * Returns the action type
     */
    get actionType() {
        return this.payload.action;
    }
    /**
     * Returns the event type
     */
    get eventType() {
        return this.payload.event_type;
    }
    /**
     * Returns the creation time from
     */
    get actionAt() {
        return this.payload.action_time;
    }
    /**
     * Returns the geo location
     */
    get geo() {
        return this.payload.geo;
    }
    /**
     * Returns the rule tags
     */
    get tags() {
        return this.payload.tags;
    }
    /**
     * Returns the identifier signer user
     */
    get signerId() {
        return this.payload.signer_id;
    }
    /**
     * Returns the information of author
     */
    get author() {
        return this.payload.author;
    }
    /**
     * Returns the identifier author
     */
    get authorId() {
        return this.payload.author.id;
    }
    /**
     * Returns the author url
     */
    get authorUrl() {
        return this.payload.author.author_url;
    }
    /**
     * Returns the identifier of the author of the original post
     */
    get sharedAuthorId() {
        return this.payload.author.shared_post_author_id;
    }
    /**
     * Returns the author url of the original post
     */
    get sharedAuthorUrl() {
        return this.payload.author.shared_post_author_url;
    }
    /**
     * Returns the author platform
     *
     * - `1` - m.vk.com or other unknown application
     * - `2` - iPhone
     * - `3` - iPad
     * - `4` - Android
     * - `5` - Windows Phone
     * - `6` - Windows
     * - `7` - vk.com or other unknown application
     */
    get authorPlatform() {
        return this.payload.author.platform;
    }
    /**
     * Returns the custom data
     */
    [vkIo.kSerializeData]() {
        const properties = [
            'url',
            'created',
            'text',
            'sharedText',
            'sharedAt',
            'actionType',
            'eventType',
            'actionAt',
            'geo',
            'tags',
            'signerId',
            'author',
            'authorId',
            'authorUrl',
            'sharedAuthorId',
            'sharedAuthorUrl',
            'authorPlatform',
            'isNew',
            'isUpdate',
            'isDelete',
            'isRestore',
            'isPost',
            'isShare',
            'isComment'
        ];
        const filtredEmptyProperties = properties.filter(property => (this[property] !== undefined));
        return copyParams(this, filtredEmptyProperties);
    }
}
vkIo.applyMixins(StreamingContext, [
    vkIo.Attachmentable
]);

const debug = createDebug__default['default']('vk-io:streaming');
class StreamingAPI {
    /**
     * Constructor
     */
    constructor({ api, updates, ...options }) {
        this.started = false;
        this.api = api;
        this.updates = updates;
        this.options = {
            agent: https.globalAgent,
            ...options
        };
    }
    /**
     * Returns custom tag
     */
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    /**
     * Starts websocket
     */
    async startWebSocket() {
        this.started = true;
        try {
            const { key, endpoint } = await this.api.streaming.getServerUrl({});
            this.key = key;
            this.endpoint = new url.URL(`https://${endpoint}`);
            const search = new url.URLSearchParams({ key });
            const { agent } = this.options;
            this.socket = new WebSocket__default['default'](`wss://${endpoint}/stream?${search}`, { agent });
        }
        catch (error) {
            this.started = false;
            throw error;
        }
        const { socket } = this;
        socket.on('message', async (data) => {
            let message;
            try {
                message = JSON.parse(data);
            }
            catch (error) {
                debug('JSON parse failed', error);
                return;
            }
            const { code } = message;
            try {
                switch (code) {
                    case 100: {
                        await this.handleEvent(message.event);
                        break;
                    }
                    case 300: {
                        await this.handleServiceMessage(message.service_message);
                        break;
                    }
                    default: {
                        debug(`Unsupported message code: ${code}`);
                    }
                }
            }
            catch (error) {
                // eslint-disable-next-line no-console
                console.log('Handle event error', error);
            }
        });
        socket.on('error', (error) => {
            debug('WebSocket error', error);
        });
    }
    /**
     * Stop all connection
     */
    async stop() {
        if (!this.started) {
            return;
        }
        this.socket.close();
        this.started = false;
    }
    /**
     * Processes server messages
     */
    async handleServiceMessage({ service_code: code }) {
        if ([3000, 3001].includes(code)) {
            await this.stop();
            await this.startWebSocket();
        }
    }
    /**
     * Handles events
     */
    async handleEvent(event) {
        const context = new StreamingContext({
            api: this.api,
            // @ts-expect-error
            upload: this.updates.upload,
            payload: event,
            state: {},
            updateType: 'publication',
            source: vkIo.UpdateSource.WEBSOCKET
        });
        await this.updates.dispatchMiddleware(context);
    }
    /**
     * Executes the HTTP request for rules
     */
    async fetchRules(method, payload = {}) {
        const { agent } = this.options;
        const url$1 = new url.URL('/rules', this.endpoint);
        url$1.searchParams.set('key', this.key);
        let body;
        if (method !== 'GET') {
            body = JSON.stringify(payload);
        }
        const response = await fetch__default['default'](url$1, {
            agent,
            method,
            body,
            headers: {
                'content-type': 'application/json'
            }
        });
        const result = await response.json();
        if (result.error !== undefined) {
            throw new StreamingRuleError(result.error);
        }
        return result;
    }
    /**
     * Returns a list of rules
     */
    async getRules() {
        const { rules = [] } = await this.fetchRules('GET');
        return rules;
    }
    /**
     * Adds a rule
     */
    async addRule(rule) {
        await this.fetchRules('POST', { rule });
    }
    /**
     * Removes the rule
     */
    async deleteRule(tag) {
        await this.fetchRules('DELETE', { tag });
    }
    /**
     * Adds a list of rules
     */
    async addRules(rules) {
        await Promise.all(rules.map(rule => (this.addRule(rule))));
    }
    /**
     * Removes all rules
     */
    async deleteRules() {
        const rules = await this.getRules();
        await Promise.all(rules.map(({ tag }) => (this.deleteRule(tag))));
    }
}
inspectable.inspectable(StreamingAPI, {
    // @ts-expect-error
    serialize: ({ started }) => ({
        started
    })
});

exports.StreamingAPI = StreamingAPI;
exports.StreamingContext = StreamingContext;
exports.StreamingRuleError = StreamingRuleError;
